### PR TITLE
[fix] Remove offline & online handlers upon "destroy"

### PR DESCRIPTION
**Pull request by wizcorp-olivier**
This should solve #399 

One remaining question/issue is:

Currently the Primus#offline method calls Primus.end(), without checking if there is or not a pending reconnection. Since Primus#end() terminates any pending attempt to reconnect, is it the desired behavior?
We saw events in this order:
- Websocket error
- ```reconnect scheduled``` (attempt number 1)
- ```offline```
- ```end``` (caused by offline as explained above)
- ```online```

So in the end we lost the connection and did not reconnect automatically.

### CODE REVIEW

**Comment by lpinca** [comment]
> Currently the Primus#offline method calls Primus.end(), without checking if there is or not a pending reconnection. Since Primus#end() terminates any pending attempt to reconnect, is it the desired behavior?

I think it is, the connection will be opened again when the `online` event is [fired](https://github.com/primus/primus/blob/66b5272041be41562ea1af46e340b3cb98c1fcb4/primus.js#L619-L621) assuming that the `online` strategy is enabled.

File: primus.js [hunk 1]
```diff
@@ -621,15 +621,29 @@ Primus.prototype.initialise = function initialise(options) {
...
   } else if (document.body.attachEvent){
-    document.body.attachEvent('onoffline', offline);
-    document.body.attachEvent('ononline', online);
+    document.body.attachEvent('onoffline', this.offlineHandler);
+    document.body.attachEvent('ononline', this.onlineHandler);
   }
+};
 
-  return primus;
+Primus.prototype.removeGlobalListeners = function () {
```
Comment (lpinca) [inline primus.js]: We should only run this in the browser, so if the network events are disabled we should bail out.
```js
if (!primus.NETWORK_EVENTS) return primus;
```

**Comment by lpinca** [comment]
This LGTM but there is one thing I'm not sure about.
Assume that the developer calls `primus.end()` and an `online` event if fired after that. Do we want to reopen the connection?

File: primus.js [hunk 2]
```diff
@@ -621,15 +621,29 @@ Primus.prototype.initialise = function initialise(options) {
     }
   }
 
+  primus.addGlobalListeners();
+
+  return primus;
+};
+
+Primus.prototype.addGlobalListeners = function () {
```
Comment (3rd-Eden) [inline primus.js]: Why is there a need for a separate function here? Can't the assignment just stay in the init method?

File: primus.js (hunk 1 above)
Comment (wizcorp-olivier) [inline primus.js]: Indeed, thanks. I could also move the 2 handler functions inside ```addGlobalListeners``` for readability - let me know if you want me to do this.

**Comment by 3rd-Eden** [comment]
@lpinca Yes, that might be an edge case, but not related to this pull request.

**Comment by 3rd-Eden** [comment]
But good solid fix for your issue, forgetting to clean up the handlers is a pretty bad mistake from our side. So a +1 on getting this in and released asap.

File: primus.js (hunk 2 above)
Comment (wizcorp-olivier) [inline primus.js]: As you prefer - I was a bit too much trying to separate things, maybe... Having all this "global" listeners logic moved into ```addGlobalListeners``` and corresponding ```removeGlobalListeners```...

Comment (3rd-Eden) [inline primus.js]: Well the initialise method already contains all the logic that should only be executed once when a new instance is created. Adding another method that should only be called once just doesn't make a lot of sense to me. I do understand that having `addGlobalListeners` makes sense if you also have a `removeGlobalListeners` but the latter method isn't strictly required either as the `destroy` function can also just run the `after` hook as a function. (see https://github.com/unshiftio/demolish for docs)

**Comment by wizcorp-olivier** [comment]
Thanks @3rd-Eden !
@lpinca the case you mention actually happened to us in the past, and we protected our code from it using a separate flag (in my previous sample with the issue) called ```wantConnection```.
The series of events was:
- (regular traffic)
- ```primus.end()``` -> connection is closed as expected
- (some time passes)
- Client's browser loses internet => ```offline```
- Client's browser gets internet again => ```online``` -> Primus tries to reconnect

This is actually why we are not using the "online" strategy of reconnection at this point.
If you want, I could propose a PR for this one too, just let me know. :cherries:

**Comment by lpinca** [comment]
A flag set when `end` is called is actually what I was thinking about, but how do you discriminate when `end` is called voluntarily or not without creating a mess?

Anyway this is not place to discuss this, let's do this in another issue or in our IRC channel.

**Comment by 3rd-Eden** [comment]
Thanks a lot for your contribution <3!

**Pull request merged**